crates/turborepo-lib/src/run/global_hash.rs

```diff
@@ -1,72 +1,172 @@
+        .iter()
+        .map(|p| root_path.anchor(p).expect("path should be from root"))
+        .collect::<Vec<_>>();
 
-    let mut matching_env_var_map = EnvironmentVariableMap::default();
-    matching_env_var_map.union(&default_env_var_map);
-    matching_env_var_map.difference(&user_env_var_set.exclusions);
+    let mut global_file_hash_map =
+        hasher.get_hashes_for_files(root_path, &global_deps_paths, false)?;
 
-    let global_hashable_env_vars = DetailedMap {
-        all: all_env_var_map,
-        by_source: BySource {
-            explicit: explicit_env_var_map,
-            matching: matching_env_var_map,
-        },
-    };
+    if !dot_env.is_empty() {
+        let system_dot_env = dot_env.iter().map(|p| p.to_anchored_system_path_buf());
+
+        let dot_env_object = hasher.hash_existing_of(root_path, system_dot_env)?;
+
+        for (key, value) in dot_env_object {
+            global_file_hash_map.insert(key, value);
+        }
+    }
+
+    let root_external_dependencies_hash = root_workspace.get_external_deps_hash();
+
+    debug!(
+        "rust external deps hash: {}",
+        root_external_dependencies_hash
+    );
 
     Ok(GlobalHashableInputs {
+        global_cache_key: GLOBAL_CACHE_KEY,
+        global_file_hash_map,
+        root_external_dependencies_hash,
+        env: global_env,
         resolved_env_vars: Some(global_hashable_env_vars),
-        ..GlobalHashableInputs::default()
+        pass_through_env: Some(global_pass_through_env),
+        env_mode,
+        framework_inference,
+        dot_env,
     })
 }
+
+impl GlobalHashableInputs {
+    pub fn calculate_global_hash_from_inputs(mut self) -> String {
+        match self.env_mode {
+            EnvMode::Infer
+                if self
+                    .pass_through_env
+                    .as_ref()
+                    .map_or(false, |env| !env.is_empty()) =>
+            {
+                self.env_mode = EnvMode::Strict;
+            }
+            EnvMode::Loose => {
+                self.pass_through_env = None;
+            }
+            // Collapse `None` and `Some([])` to `Some([])` in strict mode
+            // to match Go behavior
+            EnvMode::Strict if self.pass_through_env.is_none() => {
+                self.pass_through_env = Some(Vec::new());
+            }
+            _ => {}
+        }
+
+        self.calculate_global_hash()
+    }
+
+    fn calculate_global_hash(self) -> String {
+        let global_hashable = GlobalHashable {
+            global_cache_key: self.global_cache_key,
+            global_file_hash_map: self.global_file_hash_map,
+            root_external_dependencies_hash: self.root_external_dependencies_hash,
+            env: self.env,
+            resolved_env_vars: self
+                .resolved_env_vars
+                .map(|evm| evm.all.to_hashable())
+                .unwrap_or_default(),
+            pass_through_env: self.pass_through_env.unwrap_or_default(),
```

Comment:
I've very confused about the purpose of `calculate_global_hash_from_inputs` is if we just collapse `None` and `Some(vec![])` into the same value. I think we want this field to be `Option<Vec<String>>` as well.

Reply:
Actually I suspect it's okay to make it all `Vec<String>` because the Go hashing code will treat a nil slice as an empty slice when serializing to capnproto.

Reply:
This was all left over from how we did hashing previously, where nil vs empty slice did matter